### ISSUE TITLE
Unexpected error: Failure("Bound term variable not found (after unmangling)")

### ISSUE THREAD

**Issue by tchajed** [FStarLang/FStar#1091]
The following code results in an unexpected error:

```fstar
module Bound_termvar

module UInt = FStar.UInt
module U64 = FStar.UInt64

type t = { low: U64.t; high: U64.t }

assume val v: t -> n:nat{n < pow2 128}

let logand (a b: t) : Pure t
  (requires True)
  (ensures (fun r -> v r = UInt.logand (v a) (v b))) = a
```

output of `fstar.exe bound_termvar.v`
```
Unexpected error; please file a bug report, ideally with a minimized version of the source program that triggered the error.
Failure("Bound term variable not found (after unmangling): uu___#713759 in environment: x#713796, r#713795, b#713794, a#713793, v, __proj__Mkt__item__high, __proj__Mkt__item__low, Mkt, t, t, of_string, to_string, op_Less_Equals_Hat, op_Less_Hat, op_Greater_Equals_Hat, op_Greater_Hat, op_Equals_Hat, op_Greater_Greater_Hat, op_Less_Less_Hat, op_Bar_Hat, op_Amp_Hat, op_Hat_Hat, op_Percent_Hat, op_Slash_Hat, op_Star_Slash_Hat, op_Star_Percent_Hat, op_Star_Question_Hat, op_Star_Hat, op_Subtraction_Percent_Hat, op_Subtraction_Question_Hat, op_Subtraction_Hat, op_Plus_Percent_Hat, op_Plus_Question_Hat, op_Plus_Hat, gte_mask, eq_mask, lte, lt, gte, gt, eq, shift_left, shift_right, __uint_to_t, uint_to_t, lognot, logor, logxor, logand, rem, div_underspec, div, mul_div, mul_mod, mul_underspec, mul, sub_mod, sub_underspec, sub, add_mod, add_underspec, add, v_inj, v, t, t, __proj__Mk__item__v, uu___is_Mk, Mk, t', t', n, n, of_string, to_string, op_Less_Equals_Hat, op_Less_Hat, op_Greater_Equals_Hat, op_Greater_Hat, op_Equals_Hat, op_Greater_Greater_Hat, op_Less_Less_Hat, op_Bar_Hat, op_Amp_Hat, op_Hat_Hat, op_Percent_Hat, op_Slash_Hat, op_Star_Slash_Hat, op_Star_Percent_Hat, op_Star_Question_Hat, op_Star_Hat, op_Subtraction_Percent_Hat, op_Subtraction_Question_Hat, op_Subtraction_Hat, op_Plus_Percent_Hat, op_Plus_Question_Hat, op_Plus_Hat, gte_mask, eq_mask, lte, lt, gte, gt, eq, shift_left, shift_right, __uint_to_t, uint_to_t, lognot, logor, logxor, logand, rem, div_underspec, div, mul_div, mul_mod, mul_underspec, mul, sub_mod, sub_underspec, sub, add_mod, add_underspec, add, v_inj, v, t, t, __proj__Mk__item__v, uu___is_Mk, Mk, t', t', n, n, shift_right_value_lemma, shift_right_value_aux_3, shift_right_value_aux_2, shift_right_value_aux_1, shift_left_value_lemma, shift_left_value_aux_3, shift_left_value_aux_2, shift_left_value_aux_1, shift_right_logor_lemma, shift_left_logor_lemma, shift_right_logxor_lemma, shift_left_logxor_lemma, shift_right_logand_lemma, shift_left_logand_lemma, shift_right_lemma_2, shift_right_lemma_1, shift_left_lemma_2, shift_left_lemma_1, shift_right, shift_left, logand_mask, logor_disjoint, index_to_vec_ones, to_vec_lt_pow2, to_vec_mod_pow2, lognot_lemma_1, lognot_self, logor_lemma_2, logor_lemma_1, logor_self, logor_associative, logor_commutative, logxor_inv, xor_lemma, xor, logxor_lemma_2, logxor_lemma_1, logxor_self, logxor_associative, logxor_commutative, logand_lemma_2, logand_lemma_1, logand_self, logand_associative, logand_commutative, lognot_definition, logor_definition, logxor_definition, logand_definition, lognot, logor, logxor, logand, ones_nth_lemma, one_nth_lemma, pow2_nth_lemma, zero_nth_lemma, nth_lemma, nth, ones_from_vec_lemma, ones_to_vec_lemma, pow2_from_vec_lemma, pow2_to_vec_lemma, one_to_vec_lemma, zero_from_vec_lemma, zero_to_vec_lemma, slice_right_lemma, slice_left_lemma, append_lemma, from_vec_propriety, seq_slice_lemma, from_vec_aux, from_vec_lemma_2, from_vec_lemma_1, inverse_num_lemma, inverse_vec_lemma, inverse_aux, to_vec_lemma_2, to_vec_lemma_1, from_vec, from_vec, to_vec, to_vec, to_uint_t, lte, lt, gte, gt, eq, mod, div_underspec, div, mul_div, mul_mod, mul_underspec, mul, sub_mod, sub_underspec, sub, add_mod, add_underspec, add, decr_mod, incr_mod, decr_underspec, incr_underspec, decr, incr, ones, one, pow2_n, zero, uint_t, size, fits, min_int, max_int, pow2_values, shift_right_vec_lemma_2, shift_right_vec_lemma_1, shift_left_vec_lemma_2, shift_left_vec_lemma_1, shift_right_vec, shift_left_vec, lemma_xor_bounded, lognot_vec_definition, logor_vec_definition, logxor_vec_definition, logand_vec_definition, lognot_vec, lognot_vec, logor_vec, logor_vec, logxor_vec, logxor_vec, logand_vec, logand_vec, ones_vec, elem_vec, zero_vec, bv_t, pow2_modulo_modulo_lemma_2, pow2_modulo_modulo_lemma_1, pow2_modulo_division_lemma_2, pow2_modulo_division_lemma_1, pow2_multiplication_modulo_lemma_2, pow2_multiplication_modulo_lemma_1, pow2_multiplication_division_lemma_2, pow2_multiplication_division_lemma_1, modulo_modulo_lemma, modulo_division_lemma, modulo_division_lemma_0, lemma_mul_nat_pos_is_nat, lemma_mul_pos_pos_is_pos, division_multiplication_lemma, mod_pow2_div2, mod_mult_exact, lemma_mod_sub, modulo_addition_lemma, modulo_distributivity, division_sub_lemma, lemma_div_le, lemma_div_le_, division_addition_lemma, multiple_modulo_lemma, multiple_division_lemma, division_definition, division_definition_lemma_2, division_definition_lemma_1, division_propriety, multiplication_order_lemma, small_division_lemma_2, small_division_lemma_1, small_modulo_lemma_2, small_modulo_lemma_1, modulo_range_lemma, euclidean_division_definition, lemma_mod_mod, lemma_mod_plus_mul_distr, lemma_mod_plus_distr_l, lemma_mod_plus_injective, lemma_mod_plus_injective_1, lemma_mod_spec2, lemma_mod_spec, div_exact_r, lemma_div_exact, lemma_mul_sub_distr, lemma_mod_injective, lemma_mod_mul_distr_l, lemma_mod_mul_distr_l_0, lemma_mod_sub_1, lemma_mod_sub_0, lemma_mod_plus, lemma_mod_plus_1, lemma_mod_plus_0, lemma_eq_trans_2, lemma_div_lt, lemma_mod_lt, lemma_div_mod, modulo_lemma, multiply_fractions, pow2_minus, pow2_plus, pow2_le_compat, pow2_lt_compat, pow2_double_mult, pow2_double_sum, nat_over_pos_is_nat, pos_times_pos_is_pos, nat_times_nat_is_nat, mul_ineq1, mul_binds_tighter, swap_neg_mul, neg_mul_right, neg_mul_left, swap_mul, swap_add_plus_minus, subtraction_is_distributive, addition_is_associative, paren_add_right, paren_add_left, paren_mul_right, paren_mul_left, distributivity_sub_right, distributivity_sub_left, distributivity_add_right, distributivity_add_left, lemma_mult_lt_sqr, lemma_mult_lt_right, lemma_mult_lt_left, lemma_mult_le_right, lemma_mult_le_left, lemma_eucl_div_bound, euclidean_div_axiom, div_non_eucl_bigger_denom_lemma, div_non_eucl_decr_lemma, signed_modulo_property, abs_mul_lemma, powx_lemma2, powx_lemma1, op_Plus_Percent, signed_modulo, arithmetic_shift_right, shift_left, div_non_eucl, div, min, max, abs, powx, powx, log_2, log_2, slash_star_axiom, lemma_mul_minus_distr_l, slash_decr_axiom, mul_lemma, lemma_div_def, mem_seq_of_list, seq_of_list_tl, slice_slice, slice_length, slice_is_empty, cons_index_slice, snoc_slice_index, mem_cons, index_tail, append_cons, index_cons_r, index_cons_l, suffix_of_tail, head_cons, cons_head_tail, suffix_of, find_l_none_no_index, append_slices, append_cons_snoc, contains_cons, lemma_find_l_contains, contains_snoc, append_contains_equiv, intro_append_contains_from_disjunction, lemma_contains_singleton, lemma_contains_empty, contains_elim, contains_intro, contains, lemma_index_is_nth, createL, createL_post, lemma_list_seq_bij, lemma_seq_list_bij, seq_of_list, seq_of_list, seq_to_list, seq_to_list, seq_mem_k, seq_find, seq_find_aux, seq_find_aux, found, find_r, find_r, un_snoc, find_snoc, find_append_none_s2, find_append_none, find_append_some, find_l, find_l, lemma_mem_snoc, lemma_snoc_inj, lemma_tail_snoc, lemma_cons_snoc, snoc, lemma_trans_perm, lemma_weaken_perm_right, lemma_weaken_perm_left, lemma_trans_frame, lemma_weaken_frame_left, lemma_weaken_frame_right, lemma_tail_slice, lemma_seq_frame_lo, lemma_seq_frame_hi, lemma_swap_splice, splice_refl, splice, lemma_swap_permutes_slice, lemma_swap_slice_commute, swap_frame_hi, swap_frame_lo', swap_frame_lo, lemma_ordering_hi_cons, lemma_ordering_lo_snoc, lemma_slice_snoc, lemma_slice_cons, lemma_mem_append, cons_perm, lemma_swap_permutes, permutation, lemma_swap_permutes_aux, lemma_swap_permutes_aux_frag_eq, split_5, sorted_concat_lemma, tot_ord, total_order, lemma_count_slice, lemma_mem_count, lemma_mem_inversion, lemma_append_count_aux, lemma_append_count, sorted, sorted, lemma_tl, lemma_append_cons, upd_slice, slice_upd, lemma_slice_first_in_append, lemma_slice_append, swap, mem, count, count, split_eq, lemma_split, split, lemma_cons_inj, cons, last, lemma_tail_append, lemma_head_append, tail, head, lemma_append_inj, lemma_append_len_disj, lemma_append_inj_r, lemma_append_inj_l, indexable, init_index, init_index_aux, append_empty_r, append_empty_l, append_assoc, lemma_eq_elim, lemma_eq_refl, lemma_eq_intro, eq, eq_i, eq_i, equal, hasEq_lemma, lemma_index_slice, lemma_index_app2, lemma_index_app1, lemma_index_upd2, lemma_index_upd1, lemma_index_create, lemma_len_slice, lemma_len_append, lemma_len_upd, lemma_init_len, lemma_init_aux_len, lemma_create_len, slice, op_At_Bar, append, upd, createEmpty, exFalso0, lemma_of_list, lemma_of_list_length, of_list, init, init_aux, init_aux, create, tl, hd, cons, index, length, __proj__MkSeq__item__l, uu___is_MkSeq, MkSeq, seq, assoc_precedes, memP_precedes, precedes_append_cons_prod_r, precedes_append_cons_r, precedes_tl, strict_prefix_of_or_eq_exists_append, strict_prefix_of_exists_append, mem_strict_prefix_of, map_strict_prefix_of, strict_prefix_of_correct, strict_prefix_of_trans, strict_prefix_of_cons, strict_prefix_of_or_eq_nil, strict_prefix_of_nil, index_extensionality, index_extensionality_aux, fold_left_append_monoid, fold_left_monoid, fold_left_append, map_append, fold_left_map, fold_left_invar, assoc_mem, assoc_memP_none, assoc_memP_some, assoc_append_elim_l, assoc_append_elim_r, assoc_cons_not_eq, assoc_cons_eq, assoc_nil, noRepeats_append_intro, noRepeats_append_elim, noRepeats_cons, noRepeats_nil, memP_map_elim, memP_map_intro, memP_existsb, memP_empty, mem_memP, sortWith_sorted, append_sorted, total_order, sorted, sorted, sortWith_permutation, partition_count_forall, partition_count, partition_mem_p_forall, partition_mem_forall, partition_mem, map_lemma, rev_ind, rev'_list_ind, rev_involutive, rev'_involutive, rev_append, rev'_append, rev_rev', rev_acc_rev', rev'T, rev', rev', append_length_inv_tail, append_length_inv_head, append_inv_tail, append_inv_head, append_eq_singl, append_eq_nil, append_count_forall, append_count, append_mem_forall, append_mem, append_length, append_assoc, append_l_cons, append_cons_l, append_l_nil, append_nil_l, rev_mem, rev_acc_mem, rev_length, rev_acc_length, mem_existsb, mem_empty, strict_prefix_of, strict_prefix_of, test_sort, test_sort, sortWith, sortWith, compare_of_bool, bool_of_compare, partition_length, unzip3, unzip3, unzip, split, split, assoc, assoc, noRepeats, noRepeats, subset, subset, partition, partition, choose, choose, tryPick, tryPick, tryFind, tryFind, collect, collect, for_all, for_all, mem_filter_forall, mem_filter, filter, filter, mem_filter_spec, find, find, existsb, existsb, contains, memP, memP, mem, mem, fold_left2, fold_left2, fold_right, fold_right, fold_left, fold_left, concatMap, concatMap, mapi, mapi_init, mapi_init, map, map, flatten, flatten, op_At, append, append, rev, rev_acc, rev_acc, count, count, index, index, nth, nth, length, length, tl, tail, hd, isEmpty, excluded_middle, or_elim, ghost_lemma, impl_intro_gen, forall_impl_intro, move_requires, exists_elim, impl_intro, impl_intro_gtot, forall_to_exists_2, forall_to_exists, exists_intro, forall_intro_3, forall_intro_2, forall_intro', forall_intro, forall_intro_squash_gtot_join, forall_intro_squash_gtot, lemma_to_squash_gtot, gtot_to_lemma, lemma_forall_intro_gtot, forall_intro_gtot, arrow_to_impl, impl_to_arrow, get_forall, get_equality, get_squashed, give_witness, strong_excluded_middle, join_squash, map_squash, bind_squash, return_squash, squash_double_sum, squash_double_arrow, proof_irrelevance, give_proof, get_proof, op_Star, false_elim, ignore, __proj__Mkdtuple4__item___4, __proj__Mkdtuple4__item___3, __proj__Mkdtuple4__item___2, __proj__Mkdtuple4__item___1, uu___is_Mkdtuple4, Mkdtuple4, dtuple4, __proj__Mkdtuple3__item___3, __proj__Mkdtuple3__item___2, __proj__Mkdtuple3__item___1, uu___is_Mkdtuple3, Mkdtuple3, dtuple3, dsnd, dfst, __proj__Mktuple8__item___8, __proj__Mktuple8__item___7, __proj__Mktuple8__item___6, __proj__Mktuple8__item___5, __proj__Mktuple8__item___4, __proj__Mktuple8__item___3, __proj__Mktuple8__item___2, __proj__Mktuple8__item___1, uu___is_Mktuple8, Mktuple8, tuple8, __proj__Mktuple7__item___7, __proj__Mktuple7__item___6, __proj__Mktuple7__item___5, __proj__Mktuple7__item___4, __proj__Mktuple7__item___3, __proj__Mktuple7__item___2, __proj__Mktuple7__item___1, uu___is_Mktuple7, Mktuple7, tuple7, __proj__Mktuple6__item___6, __proj__Mktuple6__item___5, __proj__Mktuple6__item___4, __proj__Mktuple6__item___3, __proj__Mktuple6__item___2, __proj__Mktuple6__item___1, uu___is_Mktuple6, Mktuple6, tuple6, __proj__Mktuple5__item___5, __proj__Mktuple5__item___4, __proj__Mktuple5__item___3, __proj__Mktuple5__item___2, __proj__Mktuple5__item___1, uu___is_Mktuple5, Mktuple5, tuple5, __proj__Mktuple4__item___4, __proj__Mktuple4__item___3, __proj__Mktuple4__item___2, __proj__Mktuple4__item___1, uu___is_Mktuple4, Mktuple4, tuple4, __proj__Mktuple3__item___3, __proj__Mktuple3__item___2, __proj__Mktuple3__item___1, uu___is_Mktuple3, Mktuple3, tuple3, snd, fst, __proj__Mktuple2__item___2, __proj__Mktuple2__item___1, uu___is_Mktuple2, Mktuple2, tuple2, __proj__Inr__item__v, uu___is_Inr, __proj__Inl__item__v, uu___is_Inl, Inr, Inl, either, invertOption, __proj__Some__item__v, uu___is_Some, uu___is_None, Some, None, option, allow_inversion, inversion, all_trivial, all_null_wp, all_assume_p, all_assert_p, all_close_wp, all_stronger, all_if_then_else, all_bind_wp, all_return, all_ite_wp, all_wp_h, all_post_h, all_pre_h, raise, lift_div_exn, ex_trivial, ex_null_wp, ex_assume_p, ex_assert_p, ex_close_wp, ex_stronger, ex_if_then_else, ex_ite_wp, ex_bind_wp, ex_return, ex_wp, ex_post, ex_pre, ex_pre, __proj__Err__item__msg, uu___is_Err, __proj__E__item__e, uu___is_E, __proj__V__item__v, uu___is_V, Err, E, V, result, st_trivial, st_null_wp, st_assume_p, st_assert_p, st_close_wp, st_stronger, st_ite_wp, st_if_then_else, st_bind_wp, st_return, st_wp_h, st_post_h, st_pre_h, assert_norm, normalize, normalize_term, string_of_int, string_of_bool, abs, min, pow2, pow2, op_Division, op_Modulus, nonzero, nonzero, pos, pos, nat, nat, cut, _assert, admitP, unsafe_coerce, magic, admit, _assume, as_ensures, as_requires, LexTop, lex_t, lex_t, returnM, decreases, smt_pat_or, smt_pat, pattern, pattern, __proj__Cons__item__tl, __proj__Cons__item__hd, uu___is_Cons, uu___is_Nil, Cons, Nil, list, strcat, array, exn, exn, op_disEquality, op_Equality, op_LessThan, op_GreaterThanOrEqual, op_GreaterThan, op_LessThanOrEqual, op_Minus, op_Addition, op_Subtraction, op_Multiply, op_Negation, op_BarBar, op_AmpAmp, int, int, guard_free, pure_wp, pure_post, pure_pre, pure_pre, range_of, labeled, string, string, range_0, range_0, range, range, l_Exists, __proj__Mkdtuple2__item___2, __proj__Mkdtuple2__item___1, uu___is_Mkdtuple2, Mkdtuple2, dtuple2, prop, prop, l_Forall, has_type, precedes, l_not, l_iff, l_imp, l_or, __proj__Right__item___0, uu___is_Right, __proj__Left__item___0, uu___is_Left, Right, Left, c_or, l_and, __proj__And__item___1, __proj__And__item___0, uu___is_And, And, c_and, b2t, eq3, uu___is_HRefl, HRefl, h_equals, eq2, uu___is_Refl, Refl, equals, l_False, l_False, l_True, l_True, squash, unit, unit, uu___is_T, T, c_True, c_True, c_False, c_False, bool, bool, eqtype, eqtype, hasEq, cps, cps, attribute, attribute")
./bound_termvar.fst(12,39-12,44): (Error) Subtyping check failed; expected type FStar.UInt.uint_t (uu___unification_ 173637 a b r); got type (n:Prims.nat{ Prims.b2t (n < Prims.pow2 128) }) (see also /usr/local/Cellar/fstar/HEAD-9f13542_1/ulib/FStar.UInt.fst(52,28-52,36))
1 error was reported (see above)
```

Note that #933 and #1026 give different examples that result in the same error message; not sure if they're all the same bug.

I'm using a recent commit of F* on OS X 10.11:
```
$ fstar.exe --version
F* 0.9.4.3
platform=Darwin_x86_64
compiler=OCaml 4.04.1
date=2017-06-20T15:49:16+0100
commit=9f13542f9 (dirty)
```

(the working tree was dirty at compile time due to Homebrew patches)

**Comment by catalin-hritcu** [comment]
Seems like a common blow up message, maybe similar to what @cpitclaudel  reported in this comment https://github.com/FStarLang/FStar/issues/933#issuecomment-303765370
Let's use this new issue to track the problem.

**Comment by mtzguido** [comment]
A different repro:
```
$ cat B.fst 
module B

let f =
    let x = let rec __f u : unit = __f u in
    __f
    in x
$ ./bin/fstar.exe B.fst 
Unexpected error; please file a bug report, ideally with a minimized version of the source program that triggered the error.
Bound term variable not found (after unmangling): uu___#149443 in environment: false_elim, ignore, __proj__Mkdtuple4__item___4, __proj__Mkdtuple4__item___3, __proj__Mkdtuple4__item___2, __proj__Mkdtuple4__item___1, uu___is_Mkdtuple4, Mkdtuple4, dtuple4, __proj__Mkdtuple3__item___3, __proj__Mkdtuple3__item___2, __proj__Mkdtuple3__item___1, uu___is_Mkdtuple3, Mkdtuple3, dtuple3, dsnd, dfst, __proj__Inr__item__v, uu___is_Inr, __proj__Inl__item__v, uu___is_Inl, Inr, Inl, either, invertOption, allow_inversion, inversion, all_trivial, all_null_wp, all_assume_p, all_assert_p, all_close_wp, all_stronger, all_if_then_else, all_bind_wp, all_return, all_ite_wp, all_wp_h, all_post_h, all_pre_h, raise, lift_div_exn, ex_trivial, ex_null_wp, ex_assume_p, ex_assert_p, ex_close_wp, ex_stronger, ex_if_then_else, ex_ite_wp, ex_bind_wp, ex_return, ex_wp, ex_post, ex_pre, ex_pre, __proj__Err__item__msg, uu___is_Err, __proj__E__item__e, uu___is_E, __proj__V__item__v, uu___is_V, Err, E, V, result, st_trivial, st_null_wp, st_assume_p, st_assert_p, st_close_wp, st_stronger, st_ite_wp, st_if_then_else, st_bind_wp, st_return, st_wp_h, st_post_h, st_pre_h, __proj__Mktuple8__item___8, __proj__Mktuple8__item___7, __proj__Mktuple8__item___6, __proj__Mktuple8__item___5, __proj__Mktuple8__item___4, __proj__Mktuple8__item___3, __proj__Mktuple8__item___2, __proj__Mktuple8__item___1, uu___is_Mktuple8, Mktuple8, tuple8, __proj__Mktuple7__item___7, __proj__Mktuple7__item___6, __proj__Mktuple7__item___5, __proj__Mktuple7__item___4, __proj__Mktuple7__item___3, __proj__Mktuple7__item___2, __proj__Mktuple7__item___1, uu___is_Mktuple7, Mktuple7, tuple7, __proj__Mktuple6__item___6, __proj__Mktuple6__item___5, __proj__Mktuple6__item___4, __proj__Mktuple6__item___3, __proj__Mktuple6__item___2, __proj__Mktuple6__item___1, uu___is_Mktuple6, Mktuple6, tuple6, __proj__Mktuple5__item___5, __proj__Mktuple5__item___4, __proj__Mktuple5__item___3, __proj__Mktuple5__item___2, __proj__Mktuple5__item___1, uu___is_Mktuple5, Mktuple5, tuple5, __proj__Mktuple4__item___4, __proj__Mktuple4__item___3, __proj__Mktuple4__item___2, __proj__Mktuple4__item___1, uu___is_Mktuple4, Mktuple4, tuple4, __proj__Mktuple3__item___3, __proj__Mktuple3__item___2, __proj__Mktuple3__item___1, uu___is_Mktuple3, Mktuple3, tuple3, snd, fst, __proj__Mktuple2__item___2, __proj__Mktuple2__item___1, uu___is_Mktuple2, Mktuple2, tuple2, __proj__Some__item__v, uu___is_Some, uu___is_None, Some, None, option, assert_norm, normalize, normalize_term, string_of_int, string_of_bool, abs, min, pow2, pow2, op_Division, op_Modulus, nonzero, nonzero, pos, pos, nat, nat, cut, _assert, admitP, unsafe_coerce, magic, admit, _assume, as_ensures, as_requires, LexTop, lex_t, lex_t, returnM, decreases, smt_pat_or, smt_pat, pattern, pattern, __proj__Cons__item__tl, __proj__Cons__item__hd, uu___is_Cons, uu___is_Nil, Cons, Nil, list, strcat, array, exn, exn, op_disEquality, op_Equality, op_LessThan, op_GreaterThanOrEqual, op_GreaterThan, op_LessThanOrEqual, op_Minus, op_Addition, op_Subtraction, op_Multiply, op_Negation, op_BarBar, op_AmpAmp, int, int, guard_free, pure_wp, pure_post, pure_pre, pure_pre, range_of, labeled, string, string, range_0, range_0, range, range, l_Exists, __proj__Mkdtuple2__item___2, __proj__Mkdtuple2__item___1, uu___is_Mkdtuple2, Mkdtuple2, dtuple2, prop, prop, l_Forall, has_type, precedes, l_not, l_iff, l_imp, l_or, __proj__Right__item___0, uu___is_Right, __proj__Left__item___0, uu___is_Left, Right, Left, c_or, l_and, __proj__And__item___1, __proj__And__item___0, uu___is_And, And, c_and, b2t, eq3, uu___is_HRefl, HRefl, h_equals, eq2, uu___is_Refl, Refl, equals, l_False, l_False, l_True, l_True, squash, unit, unit, uu___is_T, T, c_True, c_True, c_False, c_False, bool, bool, eqtype, eqtype, hasEq, cps, cps, attribute, attribute
```
(current master, commit `3a7e06292504`, F# build)

**Comment by bichhawat** [comment]
Another example with the same error:
``
let secTry (s:string): Tot Type0 =
     match s with
     | _ -> string

type uList (s:string) = (secTry s)
let refUL (#s:string) (tl:uList s) = ST.alloc tl
``

**Comment by mtzguido** [comment]
(Surely we noticed before, but note this is an encoding bug)

Yet another example is this one, from #1066
```
module A

assume val m : Type -> Type
assume val return_m : #a:Type -> a -> m a
assume val bind_m : #a:Type -> #b:Type -> m a -> (a -> m b) -> m b
assume val push_m : #a:Type -> #b:(a -> Type) -> (x:a -> m (b x)) -> m (x:a -> b x)

val push_sum : #a:Type -> #b:(a -> Type) ->
                  dtuple2 a (fun (x:a) -> m (b x)) -> m (dtuple2 a b)
let push_sum (#a:Type) (#b:(a -> Type)) (p : (dtuple2 a (fun (x:a) -> m (b x)))) =
    match p with
    | Mkdtuple2 x y ->
        bind_m #(b x) (* #(dtuple2 a b) *) y (fun (y' : b x) ->
        return_m (* #(dtuple2 a b) *) (Mkdtuple2 x y'))
```
Notably, uncommenting any or both of the implicit argument annotations cause the example to verify properly.

**Comment by mtzguido** [comment]
I've been working on this and made some progress, but it seems nasty.

The main problem is that we're feeding VCs *with free variables* to the encoding, which blows up. (So it's not an encoding bug, it's a pervasive ill-formed-VCs bug). This happens even for some current regressions tests that **pass**, just because the encoding doesn't go into the actual problematic subterm.

There are potentially many points that introduce these bad variables. It could be some mistake about the locally nameless repr (I indeed fixed one such issue, but it's not the most common), mistaking the environments for guards (I think this happened in `tc_abs`), and, most scaringly, at generalization time.

When generalizing, we break the invariant that uvars are closed, for just a few lines, and reduce/compress them away at the end to regain the invariant. The way we do that is by replacing the `Tm_uvar` node with its solution, which (AFAIU) is mostly needed since substitutions (and hence opening/closing) *will not go* into uvars, so we cannot keep the `Tm_uvar` node.

But this is not so easy, as the uvar might have popped up in many many places, and it needs to be replaced by its solution in all of those for our trick to work. One such example is the typechecking guard after generalizing, as fixed in commit 4a1d442801dd580ddc0ff5f810d6bac2d6c25633 (which is ugly, but fixes my `B.fst` example).

That's not the only one, however, and some other guards still contain some of this offending uvars (Tej's example and Cristoph's in #1215 are of this kind). I couldn't prove it without doubt, but I feel this invariant breaking uvars might be the main cause for this.

An idea to fix this is to have two kinds of uvars, one with are affected by substitution and one which is not. We never use the first kind except for generalization. I'll investigate further and give this a try tomorrow. Curious about what @nikswamy thinks about that, too.

I'll push some progress in branch `guido_1091`, which might be rebased so don't count on it.

**Comment by mtzguido** [comment]
Seems at least `A.fst` is blowing up due to a failure in type-checking dependent matches, and not related to generalization.

**Comment by mtzguido** [comment]
While not related to generalization, it's the same problem for `A.fst`, even without generalization: uvars resolved to terms with free variables.

I made it a hard failure to assign uvars to free terms, and things started failing very early (prims). After some fixes, I now get much further but it still blows up when typechecking A.fst, seemingly because the imitation solution contains free variables (as Nik noted in #291?). At a high-level I can see the problem scope doesn't correspond to the scope that `imitation_sub_probs` gets passed as argument. I'm having a hard time making sense out of it though, since I'm not sure what a `im_or_proj_t` represents.

I'm stopping for a while to do some other stuff.

**Comment by nikswamy** [comment]
Thanks for the work on this one Guido. I'll put this one high on my list for next week.

**Comment by mtzguido** [comment]
This is now appearing in Vale, after moving from F* commit 9c3cc3626 to b53c191fb427a2d5989ce9963cd761be2f9fe3ce

**Comment by nikswamy** [comment]
If this is indeed related to generalization, what happens if we annotate the `lemma_poly_reduce` with a `val` declaration?

That should disable generalization for it.

**Comment by mtzguido** [comment]
It has a val in the interface, but the problem is not just generalization. I didn't trace this one down, but `push_sum` is also not generalized and presents the same behaviour.

**Comment by nikswamy** [comment]
I'm looking at some of the simpler examples on this thread first, starting with Abhishek's

**Comment by mtzguido** [comment]
Makes sense. You might find some commits in `guido_1091` useful. Some are fixes and some are about aborting early when things go wrong.

**Comment by nikswamy** [comment]
One aspect of the problem here, as Guido points out, is in the inference of logical guards when solving unification problems. 

It seems clear that we break the closed-term invariant there in some hard-to-diagnose places.

Now, as for fixing it, I'm a bit reluctant to dive into that code because that is precisely the code that should disappear once c_fstar-build_two_phase_checker branch lands.

In particular, with two phases, we no longer need to compute these logical guards together with solving the unification constraints. Indeed, with two phases, it should be the case that we never have to introduce these unification variables for the guards, since we will only be comparing fully elaborated types to each other.

Note, here is a minimalistic version of Abhishek's example:

```
module Bug1091
assume val preorder : Type0 -> Type0  
assume type uList (s:string) : Type0
assume val mref (a:Type0) (rel:preorder a) :Type0
assume val alloc: #a:Type -> #p:preorder a -> v:a -> mref a p
let refUL (s:string) (t:uList s) = alloc t
```
It fails in the c_fstar-build_two_phase_checker branch also, but here you can see clearly what went wrong. In the 2nd phase, it's trying to type-check the following term:

```
let  Bug1091.refUL : 
    (#uu___#156968:(Bug1091.preorder (Bug1091.uList s#156944)) //<--- clearly a variable that's not in scope
  -> s#156969:Prims.string 
  -> t#156970:(Bug1091.uList s@0:Tm_fvar: Prims.string) 
  -> Tot (Bug1091.mref (Bug1091.uList s@1:Tm_fvar: Prims.string) uu___@2:Tm_app)) = 
  (fun #uu___#156971 -> ((fun s#156972 t#156973 -> ((Bug1091.alloc #(Bug1091.uList s@1:Tm_fvar: Prims.string) #uu___@2:Tm_app t@0:Tm_delayed) $$ (residual) Prims.Tot (Bug1091.mref (Bug1091.uList s@1:Tm_fvar: Prims.string) uu___@2:Tm_app))) $$ (residual) Prims.Tot (s#156974:Prims.string -> t#156975:(Bug1091.uList s@0:Tm_fvar: Prims.string) -> Tot (Bug1091.mref (Bug1091.uList s@1:Tm_fvar: Prims.string) uu___@2:Tm_app))))
in
()
```

This second phase is going to catch many such bugs for us. @aseemr what's prognosis for getting this to master?

**Comment by aseemr** [comment]
@nikswamy, working on it, will try to make it usable asap.

To clarify, the current 2-phase mechanism would catch this sort of bug, but the fix would come later when we actually simplify the compiler code to disentangle unification and logical guards etc.

**Comment by nikswamy** [comment]
Here's another one from @karthikbhargavan  this morning:

```
module Spec.Lib.IntBuf
open FStar.HyperStack
open FStar.HyperStack.ST
let deref m a = sel m a
```

**Comment by nikswamy** [comment]
I have a partial fix in `nik_1091` that I'm trying to merge into master now.

At least part of the problem is in generalization, where we were incorrectly generalizing variable whose type may not be closed. 

I see generalization as being there mainly in support of ML-style polymorphism, which allows implicitly generalizing over variable of type `Type` only, not variables of arbitrary type. I've put in a restriction now that allows generalization only over variables whose type is a closed refinement of `Type`, i.e., impliict generalization over `Type, `eqtype` etc. is supported, but nothing else.

This fixes Tej's original report; Abhishek's report; and Karthik's report.

But, it doesn't fix Guido's module `B` and `A`.

**Comment by mtzguido** [comment]
I managed to fix `B.fst` by also removing uvars in the generated guard, and also found a couple bugs in Rel. Nik: should I merge those (as time permits)? Maybe we don't care about the first one due to moving to 2 phases, but the other are needed, I think.

Further I think it would be good to show big fat warnings when some invariants are broken. Maybe even have a `--defensive` flag, which fails loud and early when things go eerie (such as having open terms). I can take care of adding it if we agree on it.

**Comment by nikswamy** [comment]
Merging those changes sounds good, Guido.  And so does `--defensive`.

Also, @aseemr, as you say, the 2-phase merge will mainly detect bad inference results in its current forms. However, it may also fix (aka mask) some problems in type inference related to the inference of guards. Since, the 2-phase checker will effectively throw away the guard generated at the first phase and recompute guards once a program has been elaborated with all implicits and type annotations filled in.

**Comment by nikswamy** [comment]
FYI, from CHANGES.md:


* A revision to implicit generalization of types
    (Since [commit FStar@e15c2fa4eb4cd7a10dd74fc5532b6cac8e23b4f1])

  F* has for a while supported implicit generalization of types in
  support of ML-style, let-polymorphism. E.g., `let id x = x` would
  have the implicitly generalized type `#a:Type -> a -> Tot a`.

  However, F* would incorrectly also allow implicitly generalizing
  over variables of types other than `Type`. For example, this program
  would type-check, at a rather counter-intuitive type.

  ```
  type empty (x:False) =
  let rec t (#x:False) (n:nat) : empty x = t #x n
  let f () = t 0
  ```

  where, `f` would be given the type `#x:False -> unit -> empty x`.

  Worse, sometimes F* would generalize over types with free variables,
  leading to crashes as in bug #1091.

  We now restrict implicit generalization to variables whose type is a
  closed refinement of `Type`, e.g.,
    `let id x = x` has the same type as before;
    `let eq = op_Equality` has the type `#a:eqtype -> a -> a -> bool`;
     etc.

  This restriction is a breaking change. For a sampling of the changes
  needed to accommodate it see:
  
       [commit mitls/hacl-star@c93dd40b89263056c6dec8c606eebd885ba2984e]
       [commit FStar@8529b03e30e8dd77cd181256f0ec3473f8cd68bf]

**Comment by nikswamy** [comment]
Here's another case of it:
mitls-fstar@df553757de900044c22e302b41d20c5aff554cb9
make -C src/tls IK.fst

The bug goes away with this patch:


```
$ git diff
diff --git a/src/tls/IK.fst b/src/tls/IK.fst
index 041765f..937cc37 100644
--- a/src/tls/IK.fst
+++ b/src/tls/IK.fst
@@ -270,17 +270,23 @@ val derive:
   ST (derived_key u i v a)
   (requires fun h0 -> True)
   (ensures fun h0 r h1 -> True)
-
+
+let secret_ideal
+  (#u: usage info ii)
+  (#i: ii.t)
+  (#a: info)
+  (t: secret u i a{ ii.honest i}): table u i a = t
+
 let derive  #u #i #a k v =
   let (| pkg, derived_alg |) = u v in
   let a' = derived_alg a in
   if ii.honest (*safe*) i
   then
-    match MonotoneMap.lookup k v with
+    match MonotoneMap.lookup (secret_ideal k) v with
     | Some dk -> dk
     | None ->
       let dk = pkg.create (Derived i v) a' in
-      MonotoneMap.extend k v dk;
+      MonotoneMap.extend (secret_ideal k) v dk;
       dk
   else
     let raw =
```

The problem seems to be that implicit arguments for the applications for MonotoneMap.lookup and extend are not properly inferred and then it blows up. Indeed, in this case, we should not expect them to be inferred without something like my `secret_ideal` annotation; but, it should of course fail properly rather than exploding.

**Comment by mtzguido** [comment]
I ended up adding the `--defensive` option, tri-valued with either "no", "warn" or "fail". The default is "no", which doesn't perform any check, since they take noticeable time (CI would run in about 30mins).

It currently just checks that some terms are closed at various points. Running CI with `--defensive warn` shows hundreds of warnings, some of which seem relatively minimal, so there's some more data we could use.

**Comment by aseemr** [comment]
Are there known small repros for this bug (@protz, is there a small repro for the recent manifestation of it that you saw)?. I tried the examples in the threads above, and they all seem to be either fixed, or giving different errors.

I would basically like to see if 2-phases actually helps in some of those cases.

**Comment by mtzguido** [comment]
Here's a relatively minimal one of a uvar getting set to an open term. It doesn't cause the error during encoding, but I think it's just getting lucky.

```f#
module FStar.Algebra.Monoid

unopteq
type monoid (m:Type) =
  | Monoid :
    unit:m ->
    mult:(m -> m -> m) ->
    monoid m


let intro_monoid (m:Type) (u:m) (mult:m -> m -> m)
  : Pure (monoid m)
    (requires True)
    (ensures (fun mm -> Monoid?.unit mm == u /\ Monoid?.mult mm == mult))
=
  Monoid u mult

let lift_monoid_option (#a:Type) (m:monoid a) : monoid (option a) =
  let mult (x y:option a) =
    match x, y with
    | Some x0, Some y0 -> Some (m.mult x0 y0)
    | _, _ -> None
  in
  intro_monoid (option a) (Some m.unit) mult
```
Running with `--defensive warn` gives:
```
ulib/FStar.Algebra.Monoid.fst(21,15-21,22): (Warning 289) Internal: term is not closed (solve_prob').
t = (fun a m x y x0 y0 -> (*?u65*) _ a m x y /\ Prims.l_True)
FVs = ((y:FStar.Pervasives.Native.option a), (x:FStar.Pervasives.Native.option a), (m:FStar.Algebra.Monoid.monoid a), (a:Type))

ulib/FStar.Algebra.Monoid.fst(22,14-22,18): (Warning 289) Internal: term is not closed (solve_prob').
t = (fun a m x y uu___2832 uu___2833 uu___2834 -> (*?u182*) _ a m x y /\ Prims.l_True)
FVs = ((y:FStar.Pervasives.Native.option a), (x:FStar.Pervasives.Native.option a), (m:FStar.Algebra.Monoid.monoid a), (a:Type))

Verified module: FStar.Algebra.Monoid (275 milliseconds)
All verification conditions discharged successfully
```
The two phase checker doesn't seem to make a difference.

**Comment by mtzguido** [comment]
A smaller one:
```f#
module A

let f : option unit -> option unit  = 
  let aux =
    function
    | Some () -> Some ()
    | x -> None
  in
  aux
```
```

A.fst(7,11-7,15): (Warning 289) Internal: term is not closed (solve_prob').
t = (fun uu___12 x uu___111 -> (*?u1568*) _ uu___12 /\ Prims.l_True)
FVs = ((uu___12:FStar.Pervasives.Native.option Prims.unit))

Verified module: A (353 milliseconds)
All verification conditions discharged successfully
```
It seems to stem from the variables in the patterns, if the last one is changed from `x` to `None`, we don't get the warning. Variable `uu___12` is being abstracted but with a different index, so this might be that we're not closing the terms properly. I can take a look later.

**Comment by dranov** [comment]
I get the same error on the following code. I was originally trying to prove `mem x xs <==> mem x (reverse xs)`, and managed to minimize the code to this while still getting the error:

```fsharp
module Bug
open FStar.Tactics

val mem: #a:eqtype -> a -> list a -> Tot bool
let rec mem #a x xs =
	match xs with
	| [] -> false
	| hd :: tl -> if x = hd then true else mem x tl

let mem_sanity #a x xs =
	assert_by_tactic (mem x xs <==> mem x xs) idtac
```

**Comment by mtzguido** [comment]
Thanks @dranov, the repro is appreciated. Here's another one from @cpitclaudel:
```
let rec ackermann m n =
  if m=0 then n + 1
  else if n = 0 then ackermann (m - 1) 1
  else ackermann (m - 1) (ackermann m (n - 1))
```

**Comment by nikswamy** [comment]
At least `ackermann` is fixed in my branch `nik_rel`.

It fails rightfully with

```
ack.fst(14,39-14,40): (Error 19) Subtyping check failed; expected type n: Prims.int{%[ m - 1; n ] << %[ m; n ]}; got type Prims.int (see also ack.fst(13,2-15,46))
```
Since F* can only infer that `m` and `n` have type `int`, and the ordering `<` on ints is not well-founded.

This variant succeeds:

```
let rec ackermann (m:nat) (n:nat) =
  if m=0 then n + 1
  else if n = 0 then ackermann (m - 1) 1
  else ackermann (m - 1) (ackermann m (n - 1))
```

**Comment by mtzguido** [comment]
All these repros work and, as far as I know, we haven't observed this in a long time. Closing..
